File: apps/ios/0dteTrader/Core/Storage/ServerConfigStore.swift
import Foundation

/// Runtime server selection: a `UserDefaults` override of the build-time API
/// base URL, so self-hosters can point the app at their own backend (#59).
/// Mirrors `apps/desktop/src/core/api/ServerConfigStore.ts`.
final class ServerConfigStore: ObservableObject {
    static let storageKey = "serverBaseURL"

    /// The active base URL: the stored override if valid, else the build-time default.
    @Published private(set) var baseURL: URL

    private let defaults: UserDefaults
    private let defaultBaseURL: URL

    init(
        defaults: UserDefaults = .standard,
        defaultBaseURL: URL = AppEnvironment.current.apiBaseURL
    ) {
        self.defaults = defaults
        self.defaultBaseURL = defaultBaseURL
        self.baseURL = Self.storedBaseURL(in: defaults) ?? defaultBaseURL
    }

    /// The stored override if valid, else the build-time default.
    func load() -> URL {
        Self.storedBaseURL(in: defaults) ?? defaultBaseURL
    }

    /// Normalizes, validates, and persists a new base URL. Throws on invalid input.
    @discardableResult
    func save(_ input: String) throws -> URL {
        let url = try Self.normalize(input)
        defaults.set(url.absoluteString, forKey: Self.storageKey)
        baseURL = url
        return url
    }

    /// Clears the override, reverting to the build-time default.
    func reset() {
        defaults.removeObject(forKey: Self.storageKey)
        baseURL = defaultBaseURL
    }

    // MARK: - Normalization

    /// Trims, strips trailing `/` and a pasted `/v1` (or `/v1/health`) suffix,
    /// and validates a bare http(s) origin. Anything else — a leftover path,
    /// query, fragment, or embedded credentials — is rejected, because the
    /// stream derivation replaces the path with `/v1/stream` and would
    /// silently drop it. The result is rebuilt from scheme+host+port only.
    static func normalize(_ input: String) throws -> URL {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let components = URLComponents(string: trimmed),
              let scheme = components.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              let host = components.host, !host.isEmpty,
              components.user == nil, components.password == nil,
              components.query == nil, components.fragment == nil
        else {
            throw ServerConfigError.invalidURL
        }

        var path = components.path
        while path.hasSuffix("/") { path.removeLast() }
        if path.hasSuffix("/v1/health") {
            path.removeLast("/v1/health".count)
        } else if path.hasSuffix("/v1") {
            path.removeLast("/v1".count)
        }
        guard path.isEmpty else {
            throw ServerConfigError.invalidURL
        }

        var origin = URLComponents()
        origin.scheme = scheme
        origin.host = host
        origin.port = components.port
        guard let url = origin.url else {
            throw ServerConfigError.invalidURL
        }
        return url
    }

    /// WebSocket stream URL for any API base: `http → ws`, `https → wss`,
    /// path `/v1/stream`.
    static func streamURL(for baseURL: URL) -> URL {
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
        components.scheme = baseURL.scheme == "https" ? "wss" : "ws"
        components.path = "/v1/stream"
        return components.url!
    }

    // MARK: - Health check

    struct HealthCheckResult: Equatable {
        let ok: Bool
        let message: String
    }

    /// Probes `<baseURL>/v1/health` so the user can verify a server before saving.
    static func checkHealth(
        of input: String,
        timeout: TimeInterval = 4,
        urlSession: URLSession = .shared
    ) async -> HealthCheckResult {
        let url: URL
        do {
            url = try normalize(input).appendingPathComponent("v1/health")
        } catch {
            return HealthCheckResult(ok: false, message: error.localizedDescription)
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = timeout
        do {
            let (_, response) = try await urlSession.data(for: request)
            if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
                return HealthCheckResult(ok: false, message: "Server responded with HTTP \(http.statusCode)")
            }
            return HealthCheckResult(ok: true, message: "Server reachable, API ok")
        } catch let error as URLError where error.code == .timedOut {
            return HealthCheckResult(ok: false, message: "Timed out after \(Int(timeout))s")
        } catch {
            return HealthCheckResult(ok: false, message: "Server unreachable — check the URL")
        }
    }

    // MARK: - Private

    private static func storedBaseURL(in defaults: UserDefaults) -> URL? {
        defaults.string(forKey: storageKey).flatMap { try? normalize($0) }
    }
}

enum ServerConfigError: LocalizedError {
    case invalidURL

    var errorDescription: String? {
        "Enter a valid http(s) URL, e.g. https://your-api.up.railway.app"
    }
}

File: apps/desktop/src/core/api/ServerConfigStore.ts
import { DEFAULT_API_BASE_URL } from '../../app/config';
import { Store } from '../../core/observable';

export const SERVER_BASE_URL_KEY = 'serverBaseUrl';

/** Runtime server selection: a localStorage override of the build-time API
    base URL, so self-hosters can point the app at their own backend. */
export class ServerConfigStore extends Store<{ baseUrl: string }> {
  constructor() {
    super({ baseUrl: loadStoredBaseUrl() });
  }

  /** The stored override if valid, else the build-time default. */
  load(): string {
    return loadStoredBaseUrl();
  }

  /** Normalizes, validates, and persists a new base URL. Throws on invalid input. */
  save(input: string): string {
    const baseUrl = normalizeServerUrl(input);
    localStorage.setItem(SERVER_BASE_URL_KEY, baseUrl);
    this.set({ baseUrl });
    return baseUrl;
  }

  /** Clears the override, reverting to the build-time default. */
  reset(): void {
    localStorage.removeItem(SERVER_BASE_URL_KEY);
    this.set({ baseUrl: DEFAULT_API_BASE_URL });
  }
}

function loadStoredBaseUrl(): string {
  const stored = localStorage.getItem(SERVER_BASE_URL_KEY);
  if (stored === null) return DEFAULT_API_BASE_URL;
  // Same origin-only check as save(), so the invariant holds even for values
  // persisted before it was enforced.
  try {
    return normalizeServerUrl(stored);
  } catch {
    return DEFAULT_API_BASE_URL;
  }
}

/**
 * Trims, strips a pasted `/v1` or `/v1/health` suffix, and validates that the
 * result is a bare http(s) origin. Path-bearing bases are rejected outright:
 * `deriveStreamUrl` replaces the path, so REST would work while the WebSocket
 * silently broke.
 */
function normalizeServerUrl(input: string): string {
  const trimmed = input
    .trim()
    .replace(/\/+$/, '')
    .replace(/\/v1(\/health)?$/i, '');
  let url: URL;
  try {
    url = new URL(trimmed);
  } catch {
    throw new Error('Enter a valid http(s) URL, e.g. https://your-api.up.railway.app');
  }
  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    throw new Error('Enter a valid http(s) URL, e.g. https://your-api.up.railway.app');
  }
  if (url.username || url.password || url.search || url.hash || url.pathname !== '/') {
    throw new Error(
      'Use just the server origin — no path, query, or credentials (e.g. https://your-api.up.railway.app)',
    );
  }
  return url.origin;
}
export interface HealthCheckResult {
  ok: boolean;
  message: string;
  baseUrl: string;
}

/** Probes `<baseUrl>/v1/health` so the user can verify a server before saving. */
export async function checkServerHealth(
  baseUrl: string,
  timeoutMs = 4000,
): Promise<HealthCheckResult> {
  let normalizedBaseUrl: string;
  try {
    normalizedBaseUrl = normalizeServerUrl(baseUrl);
  } catch (error) {
    return {
      ok: false,
      message: error instanceof Error ? error.message : String(error),
      baseUrl: baseUrl.trim(),
    };
  }
  const url = `${normalizedBaseUrl}/v1/health`;
  try {
    const response = await fetch(url, { signal: AbortSignal.timeout(timeoutMs) });
    if (!response.ok) {
      return {
        ok: false,
        message: `Server responded with HTTP ${response.status}`,
        baseUrl: normalizedBaseUrl,
      };
    }
    return { ok: true, message: 'Server reachable, API ok', baseUrl: normalizedBaseUrl };
  } catch (error) {
    if (error instanceof DOMException && error.name === 'TimeoutError') {
      return {
        ok: false,
        message: `Timed out after ${Math.round(timeoutMs / 1000)}s`,
        baseUrl: normalizedBaseUrl,
      };
    }
    return { ok: false, message: 'Server unreachable — check the URL', baseUrl: normalizedBaseUrl };
  }
}
